Crashed when 0 byte free

### Describe the bug

Crashed when 0 byte storage is free, which made me a bit sad because I was trying to use it to free storage. But it fixed itself after I deleted one 11MB file

### To Reproduce

Open app

### Expected behavior

Works

### Device

Unihertz Jelly Max

### Android version

Android 15

### Root status

Magisk

### SD Maid variant

FOSS (F-Droid / GitHub)

### SD Maid 2/SE version

Build: `v1.7.1-rc0 (10701000) ~ 2814ab117/FOSS/RELEASE` Update history: `[1.5.7-rc0, 1.6.2-rc1, 1.6.3-rc0, 1.6.4-rc0, 1.6.5-rc0, 1.7.0-rc0, 1.7.1-rc0]` ROM: `Unihertz/Jelly_Max_EEA/Jelly_Max:14/UP1A.231005.007/V01.00.02:user/release-keys`

### Debug log

```shell
04-28 09:43:28.636 17310 17323 E AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-2
04-28 09:43:28.636 17310 17323 E AndroidRuntime: Process: eu.darken.sdmse, PID: 17310
04-28 09:43:28.636 17310 17323 E AndroidRuntime: java.io.IOException: Inoperable file: canonical[/data/data/eu.darken.sdmse/files/datastore/curriculum_vitae.preferences_pb.tmp] freeSpace[0]
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.util.-Lifecycles.attachFileSystemMessage(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:58)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.util.-Lifecycles.attachParentStacktrace(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:74)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.util.DrawableUtils.access$runFileDiagnosticsIfNotCorruption(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:161)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.DataStoreImpl$writeData$2.invokeSuspend(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:623)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.DataStoreImpl$writeData$2.invoke(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:48)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.FileStorageConnection.writeScope(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:171)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.DataStoreImpl.writeData$datastore_core(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:76)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.FileWriteScope$writeData$2.invokeSuspend(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:118)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.FileWriteScope$writeData$2.invoke(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:27)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.SingleProcessCoordinator.lock(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:90)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.RealImageLoader$enqueue$job$1.invokeSuspend(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:2574)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.RealImageLoader$enqueue$job$1.invoke(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:406)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at kotlin.collections.SetsKt.startUndspatched(Unknown Source:19)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at kotlinx.coroutines.JobKt.withContext(Unknown Source:80)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.DataStoreImpl.access$handleUpdate(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:75)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.RealImageLoader$enqueue$job$1.invokeSuspend(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:2448)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.RealImageLoader$enqueue$job$1.invoke(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:376)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.RealImageLoader$enqueue$job$1.invokeSuspend(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:2385)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:8)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at kotlinx.coroutines.DispatchedTask.run(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:114)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.work.impl.ToContinuation.run$kotlinx$coroutines$internal$LimitedDispatcher$Worker(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:6)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.work.impl.ToContinuation.run(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:30)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at kotlinx.coroutines.scheduling.TaskImpl.run(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:3)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:88)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@6379, Dispatchers.IO]
04-28 09:43:28.636 17310 17323 E AndroidRuntime: Caused by: java.io.FileNotFoundException: /data/data/eu.darken.sdmse/files/datastore/curriculum_vitae.preferences_pb.tmp: open failed: ENOSPC (No space left on device)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at libcore.io.IoBridge.open(IoBridge.java:574)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at java.io.FileOutputStream.<init>(FileOutputStream.java:259)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at java.io.FileOutputStream.<init>(FileOutputStream.java:208)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.FileWriteScope$writeData$2.invokeSuspend(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:289)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at androidx.datastore.core.FileWriteScope$writeData$2.invoke(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:61)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at coil.util.DrawableUtils.access$runFileDiagnosticsIfNotCorruption(r8-map-id-b1e61f986f0247d284fe3082682b7cc6b0b2a5f8d39d9464954b9305a62f0d08:56)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        ... 21 more
04-28 09:43:28.636 17310 17323 E AndroidRuntime: Caused by: android.system.ErrnoException: open failed: ENOSPC (No space left on device)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at libcore.io.Linux.open(Native Method)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8820)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        at libcore.io.IoBridge.open(IoBridge.java:560)
04-28 09:43:28.636 17310 17323 E AndroidRuntime:        ... 26 more
```

### Screenshots/Video

_No response_

### Additional context

_No response_

--- a/app-common-stats/src/main/java/eu/darken/sdmse/stats/core/TaskStatsCoordinator.kt
+++ b/app-common-stats/src/main/java/eu/darken/sdmse/stats/core/TaskStatsCoordinator.kt
@@ -6,6 +6,7 @@ import eu.darken.sdmse.common.debug.logging.asLog
 import eu.darken.sdmse.common.debug.logging.log
 import eu.darken.sdmse.common.debug.logging.logTag
 import eu.darken.sdmse.common.flow.withPrevious
+import eu.darken.sdmse.common.storage.StorageRescue
 import eu.darken.sdmse.main.core.taskmanager.TaskSubmitter
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.distinctUntilChanged
@@ -21,6 +22,7 @@ class TaskStatsCoordinator @Inject constructor(
     @AppScope private val appScope: CoroutineScope,
     private val taskSubmitter: TaskSubmitter,
     private val statsRepo: StatsRepo,
+    private val storageRescue: StorageRescue,
 ) {
     private val isStarted = AtomicBoolean(false)
 
@@ -56,6 +58,9 @@ class TaskStatsCoordinator @Inject constructor(
                     } catch (e: Exception) {
                         log(TAG, WARN) { "Failed to record idle snapshot: ${e.asLog()}" }
                     }
+                    // restoreIfPossible() catches all non-cancellation exceptions internally,
+                    // so a wrapper try/catch here would be dead code.
+                    storageRescue.restoreIfPossible()
                 }
             }
             .launchIn(appScope)

--- a/app-common/src/main/java/eu/darken/sdmse/common/storage/StorageRescue.kt
+++ b/app-common/src/main/java/eu/darken/sdmse/common/storage/StorageRescue.kt
@@ -1,0 +1,130 @@
+package eu.darken.sdmse.common.storage
+
+import android.content.Context
+import android.util.Log
+import dagger.hilt.android.qualifiers.ApplicationContext
+import eu.darken.sdmse.common.coroutine.DispatcherProvider
+import eu.darken.sdmse.common.debug.logging.Logging.Priority.INFO
+import eu.darken.sdmse.common.debug.logging.Logging.Priority.WARN
+import eu.darken.sdmse.common.debug.logging.asLog
+import eu.darken.sdmse.common.debug.logging.log
+import eu.darken.sdmse.common.debug.logging.logTag
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.NonCancellable
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileOutputStream
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Pre-allocates a small file (RESCUE_SIZE) in the app's no-backup data dir. When the
+ * device runs out of space, [releaseIfNeeded] frees that file early in app startup so
+ * DataStore / Room / log writes during App init have somewhere to land. After the user
+ * actually frees disk space, [restoreIfPossible] re-allocates the rescue.
+ *
+ * See issue #2401 for the original crash this guards against.
+ */
+@Singleton
+class StorageRescue @Inject constructor(
+    @ApplicationContext private val context: Context,
+    private val dispatcherProvider: DispatcherProvider,
+) {
+
+    private val baseDir: File
+        get() = context.noBackupFilesDir
+
+    private val rescueFile: File
+        get() = File(baseDir, RESCUE_FILE_NAME)
+
+    private val mutex = Mutex()
+
+    suspend fun restoreIfPossible() = withContext(dispatcherProvider.IO + NonCancellable) {
+        mutex.withLock {
+            try {
+                if (rescueFile.exists() && rescueFile.length() == RESCUE_SIZE) {
+                    log(TAG) { "restoreIfPossible(): already allocated" }
+                    return@withLock
+                }
+                if (rescueFile.exists()) {
+                    log(TAG) { "restoreIfPossible(): cleaning wrong-size leftover" }
+                    rescueFile.delete()
+                }
+
+                val usable = baseDir.usableSpace
+                if (usable < CREATE_THRESHOLD) {
+                    log(TAG) { "restoreIfPossible(): not enough headroom ($usable), skipping" }
+                    return@withLock
+                }
+
+                // Write actual bytes. RandomAccessFile.setLength() can produce sparse
+                // files on ext4/F2FS where blocks aren't reserved, defeating the rescue.
+                val buf = ByteArray(WRITE_BUFFER_SIZE)
+                FileOutputStream(rescueFile).use { out ->
+                    var written = 0L
+                    while (written < RESCUE_SIZE) {
+                        val n = minOf(buf.size.toLong(), RESCUE_SIZE - written).toInt()
+                        out.write(buf, 0, n)
+                        written += n
+                    }
+                    out.fd.sync()
+                }
+                log(TAG, INFO) { "Allocated rescue file: ${rescueFile.length()} bytes" }
+            } catch (e: CancellationException) {
+                throw e
+            } catch (e: Exception) {
+                log(TAG, WARN) { "Failed to allocate rescue file: ${e.asLog()}" }
+                try {
+                    rescueFile.delete()
+                } catch (_: Exception) {
+                }
+            }
+        }
+    }
+
+    companion object {
+        private const val STATIC_TAG = "StorageRescue"
+        private val TAG = logTag(STATIC_TAG)
+        private const val RESCUE_FILE_NAME = "storage_rescue.bin"
+        private const val RESCUE_SIZE = 2L * 1024 * 1024
+        // Release the rescue only when free space is below the rescue's own size:
+        // releasing it gives us back exactly RESCUE_SIZE bytes of headroom.
+        private const val RELEASE_THRESHOLD = RESCUE_SIZE
+        private const val CREATE_THRESHOLD = 10L * 1024 * 1024
+        private const val WRITE_BUFFER_SIZE = 64 * 1024
+
+        /**
+         * Synchronous, dependency-free, callable from [android.app.Application.attachBaseContext]
+         * BEFORE Hilt field-injection runs in `super.onCreate()`. Must not throw under any
+         * condition — this is the very first thing that runs on app launch.
+         *
+         * Hilt-injected singletons can fire DataStore/Room writes from their `init {}` blocks
+         * the moment they're constructed (e.g. RecorderModule.init reads
+         * debugSettings.recorderPath; ReportsDatabase.init runs Room queries). Doing the
+         * release inside onCreate() AFTER super.onCreate() is too late.
+         *
+         * Uses [android.util.Log] directly because the project's Logging system is
+         * installed inside onCreate() and isn't yet configured at attachBaseContext.
+         */
+        fun releaseIfNeeded(context: Context) {
+            try {
+                val baseDir = context.noBackupFilesDir
+                val rescueFile = File(baseDir, RESCUE_FILE_NAME)
+                val usable = baseDir.usableSpace
+
+                if (usable < RELEASE_THRESHOLD && rescueFile.exists()) {
+                    val deleted = rescueFile.delete()
+                    val after = baseDir.usableSpace
+                    Log.w(
+                        STATIC_TAG,
+                        "Released rescue (deleted=$deleted): usable $usable -> $after"
+                    )
+                }
+            } catch (e: Throwable) {
+                Log.e(STATIC_TAG, "releaseIfNeeded failed", e)
+            }
+        }
+    }
+}

--- a/app-common/src/test/java/eu/darken/sdmse/common/storage/StorageRescueTest.kt
+++ b/app-common/src/test/java/eu/darken/sdmse/common/storage/StorageRescueTest.kt
@@ -1,0 +1,174 @@
+package eu.darken.sdmse.common.storage
+
+import android.content.Context
+import android.util.Log
+import io.kotest.matchers.shouldBe
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.mockkStatic
+import io.mockk.spyk
+import io.mockk.unmockkStatic
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import testhelpers.BaseTest
+import testhelpers.coroutine.TestDispatcherProvider
+import java.io.File
+
+class StorageRescueTest : BaseTest() {
+
+    @TempDir
+    lateinit var tempDir: File
+
+    private lateinit var baseDir: File
+    private lateinit var rescueFile: File
+    private lateinit var context: Context
+    private val dispatcherProvider = TestDispatcherProvider()
+
+    private val rescueSize: Long = 2L * 1024 * 1024
+    private val createThreshold: Long = 10L * 1024 * 1024
+    private val releaseThreshold: Long = 2L * 1024 * 1024
+
+    @BeforeEach
+    fun setup() {
+        // releaseIfNeeded uses android.util.Log directly because it runs before
+        // the project's Logging system is installed. Stub Log in JVM tests.
+        mockkStatic(Log::class)
+        every { Log.w(any<String>(), any<String>()) } returns 0
+        every { Log.e(any<String>(), any<String>(), any()) } returns 0
+
+        baseDir = spyk(File(tempDir, "no_backup").apply { mkdirs() })
+        rescueFile = File(baseDir, "storage_rescue.bin")
+        context = mockk(relaxed = true)
+        every { context.noBackupFilesDir } returns baseDir
+    }
+
+    @AfterEach
+    fun teardown() {
+        rescueFile.delete()
+        baseDir.deleteRecursively()
+        unmockkStatic(Log::class)
+    }
+
+    private fun newRescue() = StorageRescue(context, dispatcherProvider)
+
+    // -------- restoreIfPossible --------
+
+    @Test
+    fun `restoreIfPossible allocates rescue when space is comfortable`() = runTest {
+        every { baseDir.usableSpace } returns 100L * 1024 * 1024
+
+        newRescue().restoreIfPossible()
+
+        rescueFile.exists() shouldBe true
+        rescueFile.length() shouldBe rescueSize
+    }
+
+    @Test
+    fun `restoreIfPossible no-ops when rescue already correct size`() = runTest {
+        every { baseDir.usableSpace } returns 100L * 1024 * 1024
+        rescueFile.writeBytes(ByteArray(rescueSize.toInt()))
+        val originalLastModified = rescueFile.lastModified()
+        Thread.sleep(20)
+
+        newRescue().restoreIfPossible()
+
+        rescueFile.exists() shouldBe true
+        rescueFile.length() shouldBe rescueSize
+        rescueFile.lastModified() shouldBe originalLastModified
+    }
+
+    @Test
+    fun `restoreIfPossible cleans wrong-size leftover and recreates`() = runTest {
+        every { baseDir.usableSpace } returns 100L * 1024 * 1024
+        rescueFile.writeBytes(ByteArray(123))
+
+        newRescue().restoreIfPossible()
+
+        rescueFile.exists() shouldBe true
+        rescueFile.length() shouldBe rescueSize
+    }
+
+    @Test
+    fun `restoreIfPossible skips when usable space well below CREATE_THRESHOLD`() = runTest {
+        every { baseDir.usableSpace } returns 0L
+
+        newRescue().restoreIfPossible()
+
+        rescueFile.exists() shouldBe false
+    }
+
+    @Test
+    fun `restoreIfPossible skips when usable space exactly at CREATE_THRESHOLD - 1`() = runTest {
+        every { baseDir.usableSpace } returns createThreshold - 1
+
+        newRescue().restoreIfPossible()
+
+        rescueFile.exists() shouldBe false
+    }
+
+    @Test
+    fun `restoreIfPossible allocates when usable space exactly at CREATE_THRESHOLD`() = runTest {
+        every { baseDir.usableSpace } returns createThreshold
+
+        newRescue().restoreIfPossible()
+
+        rescueFile.exists() shouldBe true
+        rescueFile.length() shouldBe rescueSize
+    }
+
+    @Test
+    fun `restoreIfPossible is idempotent under repeated invocations`() = runTest {
+        every { baseDir.usableSpace } returns 100L * 1024 * 1024
+        val rescue = newRescue()
+
+        val jobs = (1..5).map { async { rescue.restoreIfPossible() } }
+        jobs.awaitAll()
+
+        rescueFile.exists() shouldBe true
+        rescueFile.length() shouldBe rescueSize
+    }
+
+    // -------- releaseIfNeeded --------
+
+    @Test
+    fun `releaseIfNeeded deletes rescue when usable space below threshold`() {
+        every { baseDir.usableSpace } returns releaseThreshold - 1
+        rescueFile.writeBytes(ByteArray(rescueSize.toInt()))
+
+        StorageRescue.releaseIfNeeded(context)
+
+        rescueFile.exists() shouldBe false
+    }
+
+    @Test
+    fun `releaseIfNeeded keeps rescue when usable space at or above threshold`() {
+        every { baseDir.usableSpace } returns releaseThreshold
+        rescueFile.writeBytes(ByteArray(rescueSize.toInt()))
+
+        StorageRescue.releaseIfNeeded(context)
+
+        rescueFile.exists() shouldBe true
+    }
+
+    @Test
+    fun `releaseIfNeeded is a no-op when no rescue file exists`() {
+        every { baseDir.usableSpace } returns 0L
+        rescueFile.exists() shouldBe false
+
+        StorageRescue.releaseIfNeeded(context)
+
+        rescueFile.exists() shouldBe false
+    }
+
+    @Test
+    fun `releaseIfNeeded swallows exceptions and does not throw`() {
+        every { context.noBackupFilesDir } throws SecurityException("denied")
+
+        StorageRescue.releaseIfNeeded(context) // must not throw
+    }
+}

--- a/app/src/main/java/eu/darken/sdmse/App.kt
+++ b/app/src/main/java/eu/darken/sdmse/App.kt
@@ -1,6 +1,7 @@
 package eu.darken.sdmse
 
 import android.app.Application
+import android.content.Context
 import androidx.hilt.work.HiltWorkerFactory
 import androidx.work.Configuration
 import coil.Coil
@@ -24,6 +25,7 @@ import eu.darken.sdmse.common.error.installErrorDialogCustomizer
 import eu.darken.sdmse.setup.installShowSetupHint
 import eu.darken.sdmse.common.debug.memory.MemoryMonitor
 import eu.darken.sdmse.common.debug.recorder.core.RecorderModule
+import eu.darken.sdmse.common.storage.StorageRescue
 import eu.darken.sdmse.common.theming.Theming
 import eu.darken.sdmse.common.updater.UpdateService
 import eu.darken.sdmse.main.core.CurriculumVitae
@@ -57,8 +59,17 @@ open class App : Application(), Configuration.Provider {
     @Inject lateinit var shortcutManager: ShortcutManager
     @Inject lateinit var spaceMonitorControl: SpaceMonitorControl
     @Inject lateinit var taskStatsCoordinator: TaskStatsCoordinator
+    @Inject lateinit var storageRescue: StorageRescue
 
     private val logCatLogger = LogCatLogger()
+
+    override fun attachBaseContext(base: Context) {
+        super.attachBaseContext(base)
+        // Runs BEFORE Hilt field-injection in super.onCreate(): release the storage
+        // rescue early so the very first DataStore/Room writes triggered by
+        // singleton init blocks have somewhere to land. See StorageRescue / #2401.
+        StorageRescue.releaseIfNeeded(this)
+    }
 
     override fun onCreate() {
         super.onCreate()
@@ -97,6 +108,7 @@ open class App : Application(), Configuration.Provider {
         memoryMonitor.register()
 
         appScope.launch { coilTempFiles.cleanUp() }
+        appScope.launch { storageRescue.restoreIfPossible() }
         Coil.setImageLoader(imageLoaderFactory)
 
         curriculumVitae.updateAppLaunch()

